conformance/lang/expressions/let-expr/let_expr.balt

```diff
@@ -0,0 +1,593 @@
+    io:println(let int i = 2 int j = 2 in i + j); // @error missing comma token
+    io:println(let in i); // @error missing let variable declaration
+                          // @error undefined symbol 'i'
+    io:println(let i); // @error unknown type 'i'
+                       // @error missing equal token
+                       // @error missing identifier
+                       // @error missing identifier
+                       // @error missing identifier
+                       // @error missing in keyword
+}
+
+Test-Case: error
+Description: let-expr when variables are declared outside its' scope.
+Labels: let-expr, int
+
+function errorFunction() {
+    int i = 2;
+    io:println(let int i = 2 in i); // @error redeclared symbol 'i'
+}
+
+Test-Case: error
+Description: let-expr when the type of let-var-decl does not match
+Labels: let-expr
+
+function errorFunction() {
+    io:println(let string i = "2", int j = 3 in i + j); // @error operator '+' not defined for 'string' and 'int'
+    io:println(let int i = 2 in foo(i)); // @error incompatible types: expected 'string?', found 'int'
+    io:println(let var i = 2 in foo(i)); // @error incompatible types: expected 'string?', found 'int'
+}
+
+function foo(string? s) returns string? {
+    return s;
+}
+
+Test-Case: output
+Description: Test let-expr being an isolated expression, when the expression following `in` is isolated.
+             Referenced variables bound by a let-var-decls are initialized using isolated expressions.
+Labels: let-expr, function-call-expr, int
+
+isolated int x = 2;
+isolated int y = 4;
+
+isolated function bar(int i) returns int {
+    return i;
+}
+
+isolated function init() {
+    lock {
+        x = let int i = 20 in i;
+        io:println(x); // @output 20
+        x = let int i = 22 in bar(i);
+        io:println(x); // @output 22
+    }
+
+    lock {
+        int x1 = let int i = y in bar(i);
+        io:println(x1); // @output 4
+        int x2 = let int i = bar(200) in bar(i);
+        io:println(x2); // @output 200
+    }
+}
+
+Test-Case: error
+Description: Test error cases when let-expr is an isolated expression.
+             Non isolated expressions cannot be used to initialize let-var-decl
+Labels: let-expr, function-call-expr
+
+isolated int x = 2;
+int y = 2;
+
+isolated function bar(int i) returns int {
+    return i;
+}
+
+isolated function errorFunction() {
+    lock {
+        x = let int i = y in bar(i); // @error invalid access of mutable storage in an 'isolated' function
+    }
+}
+
+Test-Case: output
+Description: Test behaviour if the typed-binding-pattern uses var.
+             The type of the variable should be inferred from the precise static type of the expression following =.
+Labels: let-expr, function-call-expr, additive-expr
+
+function bar(int i) returns int {
+    return i;
+}
+
+function init() {
+    io:println(let var i = 2 in bar(i)); // @output 2
+    io:println(let var i = 2, var j = 3 in i + j); // @output 5
```

Comment:
How does this test using the precise type?

Reply:
@Maryamm, did you mean something like this?
```ballerina
type X "char_X";

function foo(X x) returns boolean {
    return x == "char_X";
}

public function main() {
    _ = let var x = "char_X" in foo(x);
}
```

Reply:
Yes.

Reply:
Added